### PR TITLE
Report MSI stages to GA

### PR DESCRIPTION
**Pull request by mdrohmann**
https://www.pivotaltracker.com/story/show/174407605

### CODE REVIEW

File: installers/msi-language/ActiveState/Tracker.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                Category = category,
+                Action = action,
+                Label = label,
+                Value = value,
+            };
+
+            eventTrackingParameters.ClientId = this._cid;
+
+            return await this._tracker.TrackAsync(eventTrackingParameters);
```
Comment (Naatan) [inline installers/msi-language/ActiveState/Tracker.cs:44]: Nit: why use async if you're going to await anyway?

File: installers/msi-language/StateDeploy/CustomAction.cs [hunk 2]
```diff
@@ -253,6 +245,35 @@ public static ActionResult InstallStateTool(Session session, out string stateToo
...
+                return ActionResult.Success;
+            }
+
+            Status.ProgressBar.StatusMessage(session, "Installing State Tool...");
+            Status.ProgressBar.Increment(session, 1);
+
+            var ret = _installStateTool(session, out stateToolPath);
+            if (ret == ActionResult.Success)
+            {
+                TrackerSingleton.Instance.TrackEventInBackground(session, "stage", "state-tool", "success");
```
Comment (Naatan) [inline installers/msi-language/StateDeploy/CustomAction.cs:270]: Let's also trigger this if it's installed but use the value "skipped".

File: installers/msi-language/StateDeploy/CustomAction.cs [hunk 3]
```diff
@@ -253,6 +245,35 @@ public static ActionResult InstallStateTool(Session session, out string stateToo
...
+            }
+            else if (ret == ActionResult.Failure)
+            {
+                TrackerSingleton.Instance.TrackEventInBackground(session, "stage", "state-tool", "failure");
```
Comment (Naatan) [inline installers/msi-language/StateDeploy/CustomAction.cs:274]: Wouldn't this also amount to a "finished" "failure"?

File: installers/msi-language/StateDeploy/CustomAction.cs [hunk 4]
```diff
@@ -386,8 +410,18 @@ public static ActionResult StateDeploy(Session session)
                         record.FormatString = String.Format("{0} failed with error:\n{1}", seq.Description, errorOutput);
 
                         MessageResult msgRes = session.Message(InstallMessage.Error | (InstallMessage)MessageBoxButtons.OK, record);
+                        if (!artifactsFetched)
+                        {
+                            TrackerSingleton.Instance.TrackEventSynchronously(session, "stage", "artifacts", "failure");
```
Comment (Naatan) [inline installers/msi-language/StateDeploy/CustomAction.cs:415]: Artifacts in this case refers to the whole deployment job. So you can call this for the entirety of the deploy process.

File: installers/msi-language/StateDeploy/CustomAction.cs [hunk 5]
```diff
@@ -398,6 +432,7 @@ public static ActionResult StateDeploy(Session session)
                 return ActionResult.Failure;
             }
 
+            TrackerSingleton.Instance.TrackEventSynchronously(session, "stage", "finished", "success");
```
Comment (Naatan) [inline installers/msi-language/StateDeploy/CustomAction.cs:435]: There's nothing else that can fail after this? Can't we track this a bit closer to the actual success state of the MSI installation?

File: installers/msi-language/StateDeploy/CustomAction.cs (hunk 3 above)
Comment (mdrohmann) [reply installers/msi-language/StateDeploy/CustomAction.cs:275]: All "finished" commands are now triggered after the finishing MSI dialogs are displayed (CustomExitDialog => success, CustomFatalError => failure).

File: installers/msi-language/StateDeploy/CustomAction.cs (hunk 5 above)
Comment (mdrohmann) [reply installers/msi-language/StateDeploy/CustomAction.cs]: All "finished" commands are now triggered after the finishing MSI dialogs are displayed (CustomExitDialog => success, CustomFatalError => failure).

File: installers/msi-language/ActiveState/Tracker.cs (hunk 1 above)
Comment (mdrohmann) [reply installers/msi-language/ActiveState/Tracker.cs:44]: The short answer: `TrackAsync` is what the C# library provides.  

The long answer: Every async function is supposed to have an `await` in C#.  The way I understand the idea behind it: You can write your control flow, as if you were writing a synchronized control flow, but things are actually running in threads and in parallel with nicer functionality to handle issues that occur with concurrency, like cancellation and callbacks.  I think every time you hit an `await` statement, the thread that you are executing in, could also change...

**Pull request merged**